sklearn/decomposition/cdnmf_fast.pyx

```diff
@@ -0,0 +1,133 @@
+# cython: cdivision=True
+# cython: boundscheck=False
+# cython: wraparound=False
+
+# Author: Mathieu Blondel, Tom Dupre la Tour for Cython version
+# License: BSD 3 clause
+
+import numpy as np
+cimport cython
+cimport numpy as np
+from libc.math cimport fabs, fmax, fmin
+
+
+def _update_cdnmf_fast(np.ndarray[double, ndim=2, mode='c'] W,
+                       np.ndarray[double, ndim=2, mode='c'] HHt,
+                       np.ndarray[double, ndim=2, mode='c'] XHt,
+                       int transpose_XHt):
+
+    cdef double violation = 0
+    cdef int n_components = W.shape[1]
+    cdef int n_samples = W.shape[0]  # n_features for H update
+    cdef double grad
+    cdef double pg
+    cdef double hess
+    cdef int i
+    cdef int t
+    cdef int r
+
+    with nogil:
+        for i in range(n_samples):
+            for t in range(n_components):
+
+                # gradient
+                # grad = GH[t, i] where GH = np.dot(HHt, H) - XHt
+                if transpose_XHt:
```

Comment:
Instead of this transpose trick, did you try just not specifying C mode in the function signature?